wallet-new/test/unit/Wallet/Abstract.hs

```diff
@@ -230,3 +240,187 @@ utxoRestrictToOurs p = utxoRestrictToAddr (isJust . p)
+    k (NewPending' a) = NewPending a
+
+-- | Given a 'Set' of addresses that will represent the
+genFromBlockchain :: Hash h Addr => Set Addr -> FromPreChain h -> Gen (Inductive h Addr)
+genFromBlockchain addrs fpc = do
+    runInductiveGen fpc (genInductiveFor addrs)
+
+genInductiveFor :: Hash h Addr => Set Addr -> InductiveGen h (Inductive h Addr)
+genInductiveFor addrs = do
+    chain <- getBlockchain
+    let initialActions = chainToApplyBlocks chain
+    intersperseTransactions addrs initialActions
+
+-- | The first step in converting a 'Chain into an 'Inductive' wallet is
+-- to sequence the existing blocks using 'ApplyBlock' constructors.
+chainToApplyBlocks :: Chain h a -> [Action h a]
+chainToApplyBlocks =
+    toList . map ApplyBlock' . chainBlocks
+
+-- | Once we've created our initial @['Action' h 'Addr']@, we want to
+-- insert some 'Transaction's in appropriate locations in the list. There
+-- are some properties that the inserted events must satisfy:
+--
+-- * The transaction must be after all of the blocks that confirm inputs to
+--   the transaction.
+-- * The transaction must be before the block that confirms it, if any
+--   blocks confirm it. It is not necessary that the transaction gets
+--   confirmed eventually!
+intersperseTransactions
+    :: Hash h Addr
+    => Set Addr
+    -> [Action h Addr]
+    -> InductiveGen h (Inductive h Addr)
+intersperseTransactions addrs actions = do
+    ourTxns <- findOurTransactions addrs <$> getBlockchain
+    let allTxnCount = length ourTxns
+    -- we weight the frequency distribution such that most of the
+    -- transactions will be represented by this wallet. this can be
+    -- changed or made configurable later.
+    txnToDisperseCount <- liftGen
+        . frequency
+        . zip [1 .. allTxnCount]
+        . map pure
+        $ [1 .. allTxnCount]
+
+    txnsToDisperse <- liftGen $ subsetOf txnToDisperseCount ourTxns
+
+    chain <- getBlockchain
+    ledger <- getLedger
+
+    let txnsWithRange =
+            mapMaybe
+                (\(i, t) -> (,,) t i <$> transactionMaxIndex addrs t chain ledger)
+                txnsToDisperse
+
+    txnsWithIndex <-
+        forM txnsWithRange $ \(t, lo, hi) ->
+            (,) t <$> liftGen (choose (lo, hi + 1))
+
+    pure
+        . toInductive
+        . conssect
+        . foldr
+            (\(t, i) -> IntMap.insertWith (<>) i [NewPending' t])
+            (dissect actions)
+        $ txnsWithIndex
+
+-- | Construct an 'IntMap' consisting of the index of the element in the
+-- input list pointing to a singleton list of the element the original
+-- list.
+dissect :: [a] -> IntMap [a]
+dissect = IntMap.fromList . zip [0..] . map pure
+
+-- | Reverse the operation of 'dissect'. Given an 'IntMap' originally
+-- representing the original index in the list pointing to the list of new
+-- items at that starting index, collapse that into a single list of
+-- elements.
+conssect :: IntMap [a] -> [a]
+conssect = concatMap snd . IntMap.toList
```

Comment:
`dissect` seemed fitting -- reifying the index of the item in the list to make it easy to insert things in the right place, and then `conssect` seemed like a fun pun on that. Definitely happy to have a more reasonable name though :)

Reply:
Works for me :)